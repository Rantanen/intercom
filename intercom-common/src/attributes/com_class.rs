
use super::common::*;

use idents;
use utils;
use model;

extern crate proc_macro;
use self::proc_macro::TokenStream;
use syn::*;

/// Expands the `com_class` attribute.
///
/// The attribute expansion results in the following items:
///
/// - Virtual table offset values for the different interfaces.
/// - `IUnknown` virtual table instance.
/// - `CoClass` trait implementation.
pub fn expand_com_class(
    attr_tokens: &TokenStream,
    item_tokens: TokenStream,
) -> Result<TokenStream, model::ParseError>
{
    // Parse the attribute.
    let mut output = vec![];
    let cls = model::ComClass::parse(
            &lib_name(), &attr_tokens.to_string(), &item_tokens.to_string() )?;
    let struct_ident = cls.name();

    // IUnknown vtable match. As the primary query_interface is implemented
    // on the root IUnknown interface, the self_vtable here should already be
    // the IUnknown we need.
    let mut query_interface_match_arms = vec![
        quote!(
            ::intercom::IID_IUnknown =>
                ( &vtables._ISupportErrorInfo )
                    as *const &::intercom::ISupportErrorInfoVtbl
                    as *mut &::intercom::ISupportErrorInfoVtbl
                    as ::intercom::RawComPtr
        ),
        quote!(
            ::intercom::IID_ISupportErrorInfo =>
                ( &vtables._ISupportErrorInfo )
                    as *const &::intercom::ISupportErrorInfoVtbl
                    as *mut &::intercom::ISupportErrorInfoVtbl
                    as ::intercom::RawComPtr
        ) ];
    let mut support_error_info_match_arms = vec![] ;

    // Gather the virtual table list struct field definitions and their values.
    // The definitions are needed when we define the virtual table list struct,
    // which is different for each com_class. The values are needed when we
    // construct the virtual table list.
    //
    // The primary IUnknown virtual table _MUST_ be at the beginning of the list.
    // This is done to ensure the IUnknown pointer matches the ComBox pointer.
    // We ensure this by defining the primary IUnknown methods on the
    // ISupportErrorInfo virtual table and having that at the beginning.
    let isupporterrorinfo_ident = Ident::from( "ISupportErrorInfo".to_owned() );
    let isupporterrorinfo_vtable_instance_ident =
            idents::vtable_instance( struct_ident, isupporterrorinfo_ident );
    let mut vtable_list_field_decls = vec![
        quote!( _ISupportErrorInfo : &'static ::intercom::ISupportErrorInfoVtbl ) ];
    let mut vtable_list_field_values = vec![
        quote!( _ISupportErrorInfo : &#isupporterrorinfo_vtable_instance_ident ) ];

    // Create the vtable data for the additional interfaces.
    // The data should include the match-arms for the primary query_interface
    // and the vtable offsets used for the delegating query_interface impls.
    for itf in cls.interfaces() {

        // Various idents.
        let offset_ident = idents::vtable_offset( struct_ident, *itf );
        let iid_ident = idents::iid( *itf );
        let vtable_struct_ident = idents::vtable_struct( *itf );
        let vtable_instance_ident = idents::vtable_instance( struct_ident, *itf );

        // Store the field offset globally. We need this offset when implementing
        // the delegating query_interface methods. The only place where we know
        // the actual layout of the vtable is here. Thus we need to store this
        // offset somewhere where the com_impl's can access it.
        //
        // Rust doesn't allow pointer derefs or conversions in consts so we'll
        // use an inline fn instead. LLVM should be able to reduce this into a
        // constant expression during compilation.
        output.push( quote!(
                #[inline(always)]
                #[allow(non_snake_case)]
                fn #offset_ident() -> usize {
                    unsafe {
                        &::intercom::ComBox::< #struct_ident >::null_vtable().#itf
                                as *const _ as usize
                    }
                }
        ) );

        // Add the interface in the vtable list.
        vtable_list_field_decls.push(
                quote!( #itf : &'static #vtable_struct_ident ) );
        vtable_list_field_values.push(
                quote!( #itf : &#vtable_instance_ident ) );

        // Define the query_interface match arm for the current interface.
        // This just gets the correct interface vtable reference from the list
        // of vtables.
        query_interface_match_arms.push( quote!(
            self::#iid_ident => &vtables.#itf
                    as *const &#vtable_struct_ident
                    as *mut &#vtable_struct_ident
                    as ::intercom::RawComPtr
        ) );

        // Define the support error info match arms.
        support_error_info_match_arms.push( quote!(
            self::#iid_ident => true
        ) );

        // ComClass (which is what the struct should be constructed to)
        // can be .into()'d into ComRc and ComItf. Generate the impls for this.
        let into_expect_msg = format!(
            "query_interface( {} ) failed for {}",
            iid_ident, itf );
        output.push( quote!(
            impl From< ::intercom::ComClass< #struct_ident > > for
                    ::intercom::ComRc< #itf > {

                fn from( source : ::intercom::ComClass< #struct_ident >) -> Self
                {
                    // into ComItf will leave the ref count dangling.
                    // This means we can just attach to get a proper ComRc.
                    let itf : ::intercom::ComItf< #itf > = source.into();
                    ::intercom::ComRc::attach( itf )
                }
            }
        ) );
        output.push( quote!(
            impl From< ::intercom::ComClass< #struct_ident > > for
                    ::intercom::ComItf< #itf > {

                fn from( source : ::intercom::ComClass< #struct_ident >) -> Self
                {
                    unsafe {

                        // ComBox::query_interface is contracted to return
                        // pointer to the correct interface. We can attach
                        // safely.
                        let itf = ::intercom::ComItf::wrap(

                            // Query interface the ComBox.
                            < #struct_ident as ::intercom::CoClass >
                                ::query_interface(
                                    ::intercom::ComBox::vtable( &source ),
                                    &#iid_ident
                                ).expect( #into_expect_msg )
                        );

                        // Forget the source. We did not increment the
                        // reference count when attaching to ComRc so we must
                        // not decrement when ComClass drops.
                        std::mem::forget( source );

                        itf
                    }
                }
            }
        ) );

        // Check if the current interface is the implicit struct interface.
        if struct_ident == itf {

            // Implicit interface.
            //
            // This interface is unimplementable from Rust perspective as it
            // represents a struct instead of a trait. Deref on ComItf will
            // deref into the struct, which we'll do through the ComBox.
            //
            // ComBox already derefs into the struct so we'll just get the
            // ComBox here and deref that.
            output.push( quote!(
                impl ::std::ops::Deref for ::intercom::ComItf< #struct_ident > {
                    type Target = #struct_ident;
                    fn deref( &self ) -> &Self::Target {
                        unsafe {
                            let self_combox =
                                    ( ::intercom::ComItf::ptr( self ) as usize
                                            - #offset_ident() )
                                    as *mut ::intercom::ComBox< #struct_ident >;

                            &**self_combox
                        }
                    }
                }
            ) );
        }
    }

    /////////////////////
    // ISupportErrorInfo virtual table instance.
    //
    // The primary IUnknown virtual table is embedded in this one.
    output.push( quote!(
            #[allow(non_upper_case_globals)]
            const #isupporterrorinfo_vtable_instance_ident
                    : ::intercom::ISupportErrorInfoVtbl
                    = ::intercom::ISupportErrorInfoVtbl {
                        __base : ::intercom::IUnknownVtbl {
                            query_interface
                                : ::intercom::ComBox::< #struct_ident >
                                    ::query_interface_ptr,
                            add_ref
                                : ::intercom::ComBox::< #struct_ident >
                                    ::add_ref_ptr,
                            release
                                : ::intercom::ComBox::< #struct_ident >
                                    ::release_ptr,
                        },
                        interface_supports_error_info
                            : ::intercom::ComBox::< #struct_ident >
                                ::interface_supports_error_info_ptr,
                    };
        ) );

    // The CoClass implementation.
    //
    // Define the vtable list struct first. This lists the vtables of all the
    // interfaces that the coclass implements.

    // VTableList struct definition.
    let vtable_list_ident = idents::vtable_list( struct_ident );
    let visibility = cls.visibility();
    output.push( quote!(
            #[allow(non_snake_case)]
            #[doc(hidden)]
            #visibility struct #vtable_list_ident {
                #( #vtable_list_field_decls ),*
            }
        ) );

    // The actual CoClass implementation.
    output.push( quote!(
            impl ::intercom::CoClass for #struct_ident {
                type VTableList = #vtable_list_ident;
                fn create_vtable_list() -> Self::VTableList {
                    #vtable_list_ident {
                        #( #vtable_list_field_values ),*
                    }
                }
                fn query_interface(
                    vtables : &Self::VTableList,
                    riid : ::intercom::REFIID,
                ) -> ::intercom::ComResult< ::intercom::RawComPtr > {
                    if riid.is_null() { return Err( ::intercom::E_NOINTERFACE ) }
                    Ok( match *unsafe { &*riid } {
                        #( #query_interface_match_arms ),*,
                        _ => return Err( ::intercom::E_NOINTERFACE )
                    } )
                }

                fn interface_supports_error_info(
                    riid : ::intercom::REFIID
                ) -> bool
                {
                    match *unsafe { &*riid } {
                        #( #support_error_info_match_arms ),*,
                        _ => false
                    }
                }
            }
        ) );

    // CLSID constant for the class.
    let clsid_ident = idents::clsid( struct_ident );
    if let Some( ref guid ) = *cls.clsid() {
        let clsid_guid_tokens = utils::get_guid_tokens( guid );
        let clsid_doc = format!( "`{}` class ID.", struct_ident );
        let clsid_const = quote!(
            #[allow(non_upper_case_globals)]
            #[doc = #clsid_doc ]
            pub const #clsid_ident : ::intercom::CLSID = #clsid_guid_tokens;
        );
        output.push( clsid_const );
    }

    Ok( tokens_to_tokenstream( item_tokens, output ) )
}
