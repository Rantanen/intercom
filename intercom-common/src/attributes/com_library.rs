
use super::common::*;

use idents;
use utils;
use model;
use builtin_model;

extern crate proc_macro;
extern crate quote;
use self::proc_macro::TokenStream;

/// Expands the `com_library` attribute.
///
/// The attribute expansion results in the following items:
///
/// - `DllGetClassObject` extern function implementation.
/// - `IntercomListClassObjects` extern function implementation.
pub fn expand_com_library(
    attr_tokens: &TokenStream,
    item_tokens: TokenStream,
) -> Result<TokenStream, model::ParseError>
{
    let mut output = vec![];
    let lib = model::ComLibrary::parse( &lib_name(), &attr_tokens.to_string() )?;

    // Create the match-statmeent patterns for each supposedly visible COM class.
    let mut match_arms = vec![];
    let mut creatable_classes = vec![];
    for struct_ident in lib.coclasses() {

        // Construct the match pattern.
        let clsid_name = idents::clsid( *struct_ident );
        match_arms.push( quote!(
            self::#clsid_name =>
                Ok( ::intercom::ComBox::new(
                        #struct_ident::new()
                    ) as ::intercom::RawComPtr )
        ) );

        // Collect class identifies of classes that have a guid and
        // which others outside the library can create.
        creatable_classes.push( quote!( #clsid_name ) );
    }

    // Generate built-in type data.
    for bti in builtin_model::builtin_intercom_types( lib.name() ) {

        // CLSID
        let clsid_tokens = utils::get_guid_tokens(
                bti.class.clsid().as_ref().unwrap() );
        let clsid_doc = format!( "Built-in {} class ID.", bti.class.name() );
        let builtin_clsid = idents::clsid( bti.class.name() );
        output.push( quote!(
            #[allow(non_upper_case_globals)]
            #[doc = #clsid_doc ]
            pub const #builtin_clsid : ::intercom::CLSID = #clsid_tokens;
        ) );

        // Match arm
        let ctor = bti.ctor;
        match_arms.push( quote!(
            self::#builtin_clsid =>
                Ok( ::intercom::ComBox::new( #ctor ) as ::intercom::RawComPtr )
        ) );

        // Include also built-in classes. They have a custom CLSID in every library.
        creatable_classes.push( quote!( #builtin_clsid ) );
    }

    // Implement DllGetClassObject.
    //
    // This is more or less the only symbolic entry point that the COM
    // infrastructure uses. The COM client uses this method to acquire
    // the IClassFactory interfaces that are then used to construct the
    // actual coclasses.
    let dll_get_class_object = get_dll_get_class_object_function( &match_arms );
    output.push( dll_get_class_object );

    // Implement DllListClassObjects
    // DllListClassObjects returns all CLSIDs implemented in the crate.
    let list_class_objects = get_intercom_list_class_objects_function( &creatable_classes );
    output.push( list_class_objects );

    Ok( tokens_to_tokenstream( item_tokens, output ) )
}

fn get_dll_get_class_object_function(
    match_arms: &[quote::Tokens]
) -> quote::Tokens
{
    let calling_convetion = get_calling_convetion();
    quote!(
            #[no_mangle]
            #[allow(non_snake_case)]
            #[allow(dead_code)]
            #[doc(hidden)]
            pub unsafe extern #calling_convetion fn DllGetClassObject(
                rclsid : ::intercom::REFCLSID,
                riid : ::intercom::REFIID,
                pout : *mut ::intercom::RawComPtr
            ) -> ::intercom::HRESULT
            {
                // Create new class factory.
                // Specify a create function that is able to create all the
                // contained coclasses.
                let mut com_struct = ::intercom::ComClass::new(
                    ::intercom::ClassFactory::new( rclsid, | clsid | {
                        match *clsid {
                            #( #match_arms, )*
                            _ => Err( ::intercom::E_NOINTERFACE ),
                        }
                    } ) );
                ::intercom::ComBox::query_interface(
                        com_struct.as_mut(),
                        riid,
                        pout );

                // com_struct will drop here and decrement the referenc ecount.
                // This is okay, as the query_interface incremented it, leaving
                // it at two at this point.

                ::intercom::S_OK
            }
        )
}

fn get_intercom_list_class_objects_function(
    clsid_tokens: &[quote::Tokens]
) -> quote::Tokens
{
    let calling_convetion = get_calling_convetion();
    let token_count = clsid_tokens.len();
    quote!(
            #[no_mangle]
            #[allow(non_snake_case)]
            #[allow(dead_code)]
            #[doc(hidden)]
            pub unsafe extern #calling_convetion fn IntercomListClassObjects(
                pcount: *mut usize,
                pclsids: *mut *const ::intercom::CLSID,
            ) -> ::intercom::HRESULT
            {
                // Do not crash due to invalid parameters.
                if pcount.is_null() { return ::intercom::E_POINTER; }
                if pclsids.is_null() { return ::intercom::E_POINTER; }

                // Store the available CLSID in a static variable so that we can
                // pass them as-is to the caller.
                static AVAILABLE_CLASSES: [::intercom::CLSID; #token_count ] = [
                    #( #clsid_tokens, )*
                ];

                // com_struct will drop here and decrement the referenc ecount.
                // This is okay, as the query_interface incremented it, leaving
                // it at two at this point.
                *pcount = #token_count;
                *pclsids = AVAILABLE_CLASSES.as_ptr();

                ::intercom::S_OK
            }
        )
}
