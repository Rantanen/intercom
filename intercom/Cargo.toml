[package]
name = "intercom"
version = "0.4.0"
authors = ["Mikko Rantanen <jubjub@jubjubnest.net>"]
license = "MIT"
edition = "2018"
repository = "https://github.com/Rantanen/intercom"
description = "Utilities for writing COM visible Rust components."

[lib]
name = "intercom"

[dependencies]
intercom-attributes = { version = "0.4", path = "../intercom-attributes" }
failure = "0.1"
serde = { version = "1.0", optional = true }
serde_derive = { version = "1.0", optional = true }
handlebars = { version = "2.0", optional = true }
log = { version = "0.4" }

[dev-dependencies]
simple_logger = { version = "1.6", default-features = false }
regex = "1.5"

[target.'cfg(not(windows))'.dependencies]
libc = "0.2"

