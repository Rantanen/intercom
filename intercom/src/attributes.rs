use crate::combox::ComBoxData;
use crate::raw::RawComPtr;
use crate::type_system;
use crate::RawComResult;
use crate::REFIID;
use crate::{type_system::TypeSystem, IID};

/// Trait required by any COM coclass type.
///
/// Used to specify the virtual table for the `ComBoxData`.
pub trait ComClass
{
    type VTableList: Copy;
    const VTABLE: Self::VTableList;
    fn query_interface(vtables: &Self::VTableList, riid: REFIID) -> RawComResult<RawComPtr>;
    fn interface_supports_error_info(riid: REFIID) -> bool;
}

pub trait HasInterface<T: ComInterface + ?Sized>: ComClass {}

pub trait ComClassInterface<TInterface: ?Sized, TS: TypeSystem>: ComClass + Sized
{
    fn offset() -> usize;

    /// # Safety
    ///
    /// The `vtable` must be a valid pointer that points to the `TInterface`
    /// part of the `self`'s virtual table list.
    unsafe fn get_box<'a>(vtable: RawComPtr) -> &'a mut ComBoxData<Self>
    {
        let offset = Self::offset();
        let self_ptr = (vtable as usize - offset) as *mut _;
        &mut *self_ptr
    }
}

pub trait ComInterfaceVTableFor<I: ?Sized, S, TS: TypeSystem>: ComInterfaceVariant<TS>
{
    const VTABLE: Self::VTable;
}

/// The `ComInterface` trait defines the COM interface details for a COM
/// interface trait.
pub trait ComInterface
{
    /// The current interface.
    ///
    /// This associated type exists only to provide better error messages. When
    /// the `ComInterfaceVariant` is accessed through this type, the compiler
    /// will first report the missing `ComInterface` implementation if `Self`
    /// is not a COM interface.
    type TSelf: ?Sized;

    /// IID of the COM interface.
    fn iid(ts: type_system::TypeSystemName) -> Option<&'static IID>;

    fn iid_ts<TS: intercom::type_system::TypeSystem>() -> &'static intercom::IID
    where
        Self: intercom::attributes::ComInterfaceVariant<TS>;

    /// Dereferences a `ComItf<T>` into a `&T`.
    ///
    /// While in most cases the user crate will implement `T` for `ComItf<T>`,
    /// this impl exists only in the user crate and cannot be used in generic
    /// contexts. For generic `ComItf<T>` use, Intercom ipmls `Deref<Target=T>`
    /// for `ComItf<T>` which requires this method.
    fn deref(com_itf: &crate::ComItf<Self>) -> &Self;
}

pub trait ComInterfaceVariant<TS: TypeSystem>
{
    type VTable: Copy + 'static;
    fn iid() -> &'static IID;
}

pub trait ComClassTypeInfo
{
    fn gather_type_info() -> Vec<crate::typelib::TypeInfo>;
}

pub trait ComInterfaceTypeInfo
{
    fn gather_type_info() -> Vec<crate::typelib::TypeInfo>;
}
