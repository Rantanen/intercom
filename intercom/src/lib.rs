#![crate_type="dylib"]
#![feature(unique, shared)]

mod classfactory; pub use classfactory::*;
mod combox; pub use combox::*;
mod comrc; pub use comrc::*;
mod comitf; pub use comitf::*;
mod bstr; pub use bstr::*;
mod guid; pub use guid::GUID;
mod error; pub use error::{return_hresult, get_last_error, ComError};

// The crate doesn't really need the macros. However Rust will complain that
// the import does nothing if we don't define #[macro_use]. Once we define
// #[macro_use] to get rid of that warning, Rust will complain that the
// #[macro_use] does nothing. Fortunately THAT warning comes with a named
// warning option so we can allow that explicitly.
#[allow(unused_imports)]
#[macro_use]
extern crate intercom_attributes;
pub use intercom_attributes::*;

/// Raw COM pointer type.
pub type RawComPtr = *mut std::os::raw::c_void;

/// Interface ID GUID.
pub type IID = GUID;

/// A reference to an interface ID.
pub type REFIID = *const IID;

/// Class ID GUID.
pub type CLSID = GUID;

/// A reference to a class ID.
pub type REFCLSID = *const IID;

/// COM error result value.
pub type HRESULT = i32;

/// HRESULT indicating the operation completed successfully.
pub const S_OK : HRESULT = 0;

/// HRESULT indicating the operation completed successfully and returned FALSE.
pub const S_FALSE : HRESULT = 1;

/// HRESULT for unimplemented functionality.
#[allow(overflowing_literals)]
pub const E_NOTIMPL : HRESULT = 0x80004001 as HRESULT;

/// HRESULT indicating the type does not support the requested interface.
#[allow(overflowing_literals)]
pub const E_NOINTERFACE : HRESULT = 0x80004002 as HRESULT;

/// HRESULT indicating a pointer parameter was invalid.
#[allow(overflowing_literals)]
pub const E_POINTER : HRESULT = 0x80004003 as HRESULT;

/// HRESULT for aborted operation.
#[allow(overflowing_literals)]
pub const E_ABORT : HRESULT = 0x80004004 as HRESULT;

/// HRESULT for unspecified failure.
#[allow(overflowing_literals)]
pub const E_FAIL : HRESULT = 0x80004005 as HRESULT;

/// HRESULT for invalid argument.
#[allow(overflowing_literals)]
pub const E_INVALIDARG : HRESULT = 0x80070057 as HRESULT;

/// IUnknown interface ID.
#[allow(non_upper_case_globals)]
pub const IID_IUnknown : GUID = GUID {
    data1: 0x00000000, data2: 0x0000, data3: 0x0000,
    data4: [ 0xC0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x46 ]
};

/// IClassFactory interface ID.
#[allow(non_upper_case_globals)]
pub const IID_IClassFactory : GUID = GUID {
    data1: 0x00000001, data2: 0x0000, data3: 0x0000,
    data4: [ 0xC0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x46 ]
};

/// ISupportErrorInfo interface ID.
#[allow(non_upper_case_globals)]
pub const IID_ISupportErrorInfo : GUID = GUID {
    data1: 0xDF0B3D60, data2: 0x548F, data3: 0x101B,
    data4: [ 0x8E, 0x65, 0x08, 0x00, 0x2B, 0x2B, 0xD1, 0x19 ]
};

/// IErrorInfo interface ID.
#[allow(non_upper_case_globals)]
pub const IID_IErrorInfo : GUID = GUID {
    data1: 0x1CF2B120, data2: 0x547D, data3: 0x101B,
    data4: [ 0x8E, 0x65, 0x08, 0x00, 0x2B, 0x2B, 0xD1, 0x19 ]
};

/// IUnknown virtual table layout.
#[repr(C)]
pub struct IUnknownVtbl
{
    /// QueryInterface method pointer.
    pub query_interface : unsafe extern "stdcall" fn(
        s : RawComPtr,
        _riid : REFIID,
        out : *mut RawComPtr
    ) -> HRESULT,

    /// AddRef method pointer.
    pub add_ref: unsafe extern "stdcall" fn( s : RawComPtr ) -> u32,

    /// Release method pointer.
    pub release: unsafe extern "stdcall" fn( s : RawComPtr ) -> u32,
}

/// ISupportErrorInfo virtual table layout.
#[repr(C)]
pub struct ISupportErrorInfoVtbl
{
    /// ISupportErrorInfo inherits IUnknown.
    pub __base : IUnknownVtbl,

    /// QueryInterface method pointer.
    pub interface_supports_error_info : unsafe extern "stdcall" fn(
        s : RawComPtr,
        riid : REFIID,
    ) -> HRESULT,
}

// Do we need this? Would rather not export this through an extern crate
// for another dll.
//
// com_library should have dllmain!() macro or similar that implements this
// together with the COM registration.
#[no_mangle]
#[allow(non_camel_case_types)]
pub extern "stdcall" fn DllMain(
    _dll_instance : *mut std::os::raw::c_void,
    _reason : u32,
    _reserved : *mut std::os::raw::c_void ) -> bool
{
    true
}

pub type ComResult<A> = Result<A, HRESULT>;

