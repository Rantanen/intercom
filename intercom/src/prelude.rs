pub use crate::{
    com_class, com_interface, com_library, com_module, com_signature, ComBox, ComError, ComItf,
    ComRc, ComResult,
};
