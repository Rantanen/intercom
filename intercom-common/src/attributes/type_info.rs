use crate::prelude::*;

/// Expands the `ForeignType` derive attribute.
///
/// The attribute expansion results in the following items:
///
/// - Implementation of the ForeignType trait.
pub fn expand_bidirectional_type_info(
    item_tokens: TokenStreamNightly,
) -> Result<TokenStreamNightly, syn::Error>
{
    // Get the name of the type we want to implement the trait for.
    let input: syn::DeriveInput = syn::parse(item_tokens)?;
    let name = &input.ident;

    // Immpl requires the the generics in particular way.
    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
    let result = quote! { impl #impl_generics intercom::type_system::ForeignType for #name #ty_generics #where_clause {

        /// The default name is the name of the type.
        fn type_name() -> &'static str { stringify!( #name ) }
    } };

    Ok(result.into())
}

/// Expands the `ExternType` derive attribute.
///
/// The attribute expansion results in the following items:
///
/// - Implementation of the ExternType trait.
pub fn expand_derive_extern_type(
    item_tokens: TokenStreamNightly,
) -> Result<TokenStreamNightly, syn::Error>
{
    // Get the name of the type we want to implement the trait for.
    let input: syn::DeriveInput = syn::parse(item_tokens)?;
    let name = &input.ident;

    // Immpl requires the the generics in particular way.
    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
    let result = quote! {
        impl<TS: intercom::type_system::TypeSystem> #impl_generics intercom::type_system::ExternType<TS> for #name #ty_generics #where_clause {
            type ForeignType = #name;
        }
    };

    Ok(result.into())
}

/// Expands the `ExternInput` derive attribute.
///
/// The attribute expansion results in the following items:
///
/// - Implementation of the ExternInput trait.
pub fn expand_derive_extern_input(
    item_tokens: TokenStreamNightly,
) -> Result<TokenStreamNightly, syn::Error>
{
    // Get the name of the type we want to implement the trait for.
    let input: syn::DeriveInput = syn::parse(item_tokens)?;
    let name = &input.ident;

    // Immpl requires the the generics in particular way.
    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
    let result = quote! {
        unsafe impl<TS: intercom::type_system::TypeSystem> #impl_generics intercom::type_system::ExternInput<TS> for #name #ty_generics #where_clause {

            type Lease = ();

            #[inline(always)]
            unsafe fn into_foreign_parameter(self) -> intercom::ComResult<(Self::ForeignType, Self::Lease)> {
                Ok((self, ()))
            }

            type Owned = #name;

            #[inline(always)]
            unsafe fn from_foreign_parameter(source: Self::ForeignType) -> intercom::ComResult<Self::Owned> {
                Ok(source)
            }
        }

        unsafe impl<TS: intercom::type_system::TypeSystem> #impl_generics intercom::type_system::InfallibleExternInput<TS> for #name #ty_generics #where_clause {

            type Lease = ();

            #[inline(always)]
            unsafe fn into_foreign_parameter(self) -> (Self::ForeignType, Self::Lease) {
                (self, ())
            }

            type Owned = #name;

            #[inline(always)]
            unsafe fn from_foreign_parameter(source: Self::ForeignType) -> Self::Owned {
                source
            }
        }
    };

    Ok(result.into())
}

/// Expands the `ExternOutput` derive attribute.
///
/// The attribute expansion results in the following items:
///
/// - Implementation of the ExternOutput trait.
pub fn expand_derive_extern_output(
    item_tokens: TokenStreamNightly,
) -> Result<TokenStreamNightly, syn::Error>
{
    // Get the name of the type we want to implement the trait for.
    let input: syn::DeriveInput = syn::parse(item_tokens)?;
    let name = &input.ident;

    // Impl requires the the generics in particular way.
    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
    let result = quote! {
        unsafe impl<TS: intercom::type_system::TypeSystem> #impl_generics intercom::type_system::ExternOutput<TS> for #name #ty_generics #where_clause {

            #[inline(always)]
            fn into_foreign_output(self) -> intercom::ComResult<Self::ForeignType> {
                Ok(self)
            }

            #[inline(always)]
            unsafe fn from_foreign_output(source: Self::ForeignType) -> intercom::ComResult<Self> {
                Ok(source)
            }
        }

        unsafe impl<TS: intercom::type_system::TypeSystem> #impl_generics intercom::type_system::InfallibleExternOutput<TS> for #name #ty_generics #where_clause {

            #[inline(always)]
            fn into_foreign_output(self) -> Self::ForeignType {
                self
            }

            #[inline(always)]
            unsafe fn from_foreign_output(source: Self::ForeignType) -> Self {
                source
            }
        }
    };

    Ok(result.into())
}
