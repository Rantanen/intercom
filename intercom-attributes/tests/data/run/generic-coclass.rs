extern crate intercom;
use intercom::attributes::HasInterface;
use intercom::prelude::*;
use std::cell::RefCell;

#[com_interface]
trait IAnimal
{
    fn get_name(&self) -> ComResult<String>;
    fn describe(&self) -> ComResult<String>;
}

// Crab

#[com_interface]
trait ICrab
{
    fn pick_up(&self, item: &str) -> ComResult<()>;
}

#[com_class(IAnimal, ICrab)]
#[derive(Default)]
struct Crab
{
    name: &'static str,
    item: RefCell<Option<String>>,
}

impl Crab
{
    fn new(name: &'static str) -> Crab
    {
        Crab {
            name,
            item: RefCell::new(None),
        }
    }
}

impl ICrab for Crab
{
    fn pick_up(&self, item: &str) -> ComResult<()>
    {
        self.item.replace(Some(item.to_string()));
        Ok(())
    }
}

impl IAnimal for Crab
{
    fn get_name(&self) -> ComResult<String>
    {
        Ok(self.name.to_string())
    }

    fn describe(&self) -> ComResult<String>
    {
        match *self.item.borrow() {
            Some(ref item) => Ok(format!("A crab called {} carrying {}", self.name, item)),
            None => Ok(format!("A crab called {} who is empty pincered", self.name)),
        }
    }
}

// Dog

#[com_interface]
trait IDog
{
}

#[com_class(IAnimal, IDog)]
#[derive(Default)]
struct Dog
{
    name: &'static str,
}

impl IDog for Dog {}
impl IAnimal for Dog
{
    fn get_name(&self) -> ComResult<String>
    {
        Ok(self.name.to_string())
    }

    fn describe(&self) -> ComResult<String>
    {
        Ok(format!("A dog called {}", self.name))
    }
}

// Generic collection

#[com_interface]
trait ICollection
{
    fn get_count(&self) -> usize;
    fn get_item(&self, idx: usize) -> ComResult<ComRc<dyn IAnimal>>;
}

#[com_class(ICollection)]
struct Collection<T: HasInterface<dyn IAnimal> + Default>
{
    items: Vec<ComBox<T>>,
}

impl<T: HasInterface<dyn IAnimal> + Default> ICollection for Collection<T>
{
    fn get_count(&self) -> usize
    {
        self.items.len()
    }

    fn get_item(&self, idx: usize) -> ComResult<ComRc<dyn IAnimal>>
    {
        Ok(self
            .items
            .get(idx)
            .map(|item| item.into())
            .unwrap_or_else(|| ComBox::new(T::default()).into()))
    }
}

// A factory that returns generic collections.

#[com_class(Self)]
struct Factory;

#[com_interface]
impl Factory
{
    fn get_crabs(&self) -> ComResult<ComRc<dyn ICollection>>
    {
        let collection = ComBox::new(Collection {
            items: vec![
                ComBox::new(Crab::new("Ferris")),
                ComBox::new(Crab::new("Pincers")),
                ComBox::new(Crab::new("Shelly")),
            ],
        });

        Ok(collection.into())
    }

    fn get_dogs(&self) -> ComResult<ComRc<dyn ICollection>>
    {
        let collection = ComBox::new(Collection {
            items: vec![
                ComBox::new(Dog { name: "Spot" }),
                ComBox::new(Dog { name: "Stripe" }),
                ComBox::new(Dog { name: "Checkers" }),
            ],
        });

        Ok(collection.into())
    }
}

fn main()
{
    // Instantiate Factory as a COM interface.
    let factory = ComBox::new(Factory);
    let factory: ComRc<Factory> = factory.into();

    // Make crabs pick up stuff.
    let crab_collection = factory.get_crabs().expect("Failed to get crabs");
    let items = ["borrowck", "fish", "gear"];
    for i in 0..crab_collection.get_count() {
        let animal = crab_collection.get_item(i).expect("Failed to get item");
        let crab: ComRc<dyn ICrab> = intercom::ComItf::query_interface(&animal)
            .expect("Crab collection had a non-crab item");
        crab.pick_up(items[i]);
    }

    // Describe each crab.
    for i in 0..crab_collection.get_count() {
        let animal = crab_collection.get_item(i).expect("Failed to get item");
        println!("{}", animal.describe().expect("Failed to describe"));
    }
}
