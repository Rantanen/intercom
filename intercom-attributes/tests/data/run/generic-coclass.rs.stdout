A crab called Ferris carrying borrowck
A crab called Pincers carrying fish
A crab called Shelly carrying gear
